packages/buttons/src/elements/ButtonGroup.js

```diff
@@ -5,110 +5,85 @@
-    /**
-     * Currently selected tab to display
-     */
-    selectedKey: PropTypes.any,
-    /**
-     * @param {Object} newState
-     * @param {Any} newState.selectedKey - The newly selected key
-     */
-    onStateChange: PropTypes.func
-  };
-
-  constructor(...args) {
-    super(...args);
-
-    this.state = {
-      id: IdManager.generateId('garden-button-group'),
-      selectedKey: undefined,
-      focusedKey: undefined
-    };
-
-    this.firstKey = undefined;
-  }
-
-  componentDidMount() {
-    /**
-     * In an uncontrolled state we want to always display the first button
-     */
-    if (!this.isControlledProp('selectedKey') && typeof this.firstKey !== 'undefined') {
-      this.setControlledState({ selectedKey: this.firstKey });
+const ButtonGroup = ({ id, selectedKey, onStateChange, children, ...otherProps }) => {
+  const [controlledSelectedItem, setControlledSelectedItem] = useState(selectedKey);
+
+  const { selectedItem, focusedItem, getButtonProps, getGroupProps } = useButtonGroup({
+    id,
+    rtl: isRtl(otherProps),
+    defaultSelectedIndex: 0,
+    selectedItem: selectedKey === undefined ? controlledSelectedItem : selectedKey,
+    onSelect: newSelectedItem => {
+      if (onStateChange) {
+        onStateChange({ selectedKey: newSelectedItem });
+      } else {
+        setControlledSelectedItem(newSelectedItem);
+      }
     }
-  }
+  });
 
-  renderButtons = getButtonProps => {
-    const { children } = this.props;
-    const { selectedKey, focusedKey } = this.getControlledState();
+  const buttons = Children.map(children, child => {
+    if (!isValidElement(child)) {
+      return child;
+    }
 
-    return Children.map(children, child => {
-      if (!isValidElement(child)) {
+    if (hasType(child, Button)) {
+      if (child.props.disabled) {
         return child;
       }
 
-      if (hasType(child, Button)) {
-        if (child.props.disabled) {
-          return child;
-        }
+      const key = child.key;
 
-        const key = child.key;
-
-        if (!key) {
-          throw new Error('"key" prop must be provided to Button');
-        }
-
-        if (typeof this.firstKey === 'undefined') {
-          this.firstKey = key;
-        }
-
-        return cloneElement(
-          child,
-          getButtonProps({
-            key,
-            selected: key === selectedKey,
-            focused: key === focusedKey,
-            ...child.props
-          })
-        );
+      if (!key) {
+        throw new Error('"key" prop must be provided to Button');
       }
 
-      return child;
-    });
-  };
-
-  render() {
-    const { children, ...otherProps } = this.props;
-    const { focusedKey, selectedKey, id } = this.getControlledState();
+      const focusRef = React.createRef();
+
+      return cloneElement(
+        child,
+        getButtonProps({
+          key,
+          selected: key === selectedItem,
+          focused: key === focusedItem,
+          item: key,
+          ref: focusRef,
```

Comment:
Could you clarify the purpose of line 62? The ref value is also passed on line 63 and it seems like the value is already [defined](https://github.com/zendeskgarden/react-containers/blob/master/packages/selection/src/useSelection.js#L237) in `useSelection`. What are the consequences of removing line 62?

Reply:
Good catch, this isn't needed anymore. There was a recent change to `useSelection` that doesn't make this necessary, but this would still work.

We should make a follow-on task to update our READMEs in `react-containers` to use the correct syntax. The example in https://www.npmjs.com/package/@zendeskgarden/container-buttongroup is showing my original syntax.

Reply:
Oh nice, thanks for the clarification!